### PR TITLE
Bug 1986631: Do not drop environment variables without name but with a value, also fix crash when ref is empty

### PR DESCRIPTION
**Pull request by jerolimov**
**Fixes**: 
https://issues.redhat.com/browse/ODC-6206
https://bugzilla.redhat.com/show_bug.cgi?id=1986631

When the user makes changes to a BuildConfig (or Pod) with the environment page or tab this could result in different errors:

**Usability error:** When the user doesn't enter a name this entries are just dropped when the form is saved.

**Crash:** When the user enters a name but do not enter a ConfigMap or Secret value the form crashs after saving. (The data are saved and the form will crash now when the user opens the tab.)

**Unstable tests:** `ConfigMapSecretKeyRef` called `NameKeyDropdownPair` as a function instead of rendering it as child component. This results in a react error "Uncaught Error: Rendered more hooks than during the previous render)."

**Analysis / Root cause**: 
There are two small aligned, but separate issues: The old implementation filters all environment variables without a name and doesn't check the value.

When the secret reference was not empty, the resource doesn't contain the info which resource should be selected. `value-from-pair.jsx` crashs without this info.

Call `NameKeyDropdownPair` as jsx tag instead of a function so that this component can use a hook (useTranslation).

**Solution Description**: 
Add null check to `value-from-pair.jsx` so that the component doesn't crash when another code call this without required data. Rendering nothing here results in a broken UI:

null-check

To fix this I also updated the value or valueFrom check. The value `valueFrom: {}` breaks the UI. The new version renders now a text field for this fields:

clear-valuefrom

The change in `_envVarsToNameVal` checks now if the name AND value is empty to skip these entries. When a value is inserted by the user, this will be submitted. The server then shows an error that the name is missing. See gif below.

**Screen shots / Gifs for design review**: 
@openshift/team-devconsole-ux

Earlier the fields without a name are just dropped. Now this could be submitted to the server, but this incomplete data will then show an error:

improved-env

**Unit test coverage report**: 
Unchanged

**Test setup:**
Steps to Reproduce the usability error:
1. Open developer perspective
2. Navigate to Builds
3. Click on "Create BuildConfig" button
4. Just create the BuildConfig without additional changes
5. Switch to the "Environment" tab
6. Press "Add variable" or "Add from ConfigMap or Secret", enter an value or select a resource WITHOUT adding a name
7. Press "Save" (environment variable without a name is just dropped)

Steps to Reproduce the crash:
1. Open developer perspective
2. Navigate to Builds
3. Click on "Create BuildConfig" button
4. Just create the BuildConfig without additional changes
5. Switch to the "Environment" tab
6. Press "Add from ConfigMap or Secret", add a name but DO NOT selecting a ConfigMap or Secret
7. Press "Save"

**Browser conformance**: 
- [x] Chrome
- [ ] Firefox
- [ ] Safari
- [ ] Edge

### CODE REVIEW

**Comment by nemesis09** [comment]
looks good to me.
will add lgtm once we have approval from UX
cc @openshift/team-devconsole-ux

**Comment by bgliwa01** [comment]
lgtm!

**Review by bgliwa01 — APPROVED** [review]
lgtm!

**Comment by serenamarie125** [comment]
lgtm!

**Comment by jerolimov** [comment]
I canceled the merge above because I re-run the failing protractor test multiple times locally and noticed that doesn't run stable. Sometimes it failed with this error:

```
yarn run v1.22.10
$ yarn ts-node ./node_modules/.bin/protractor integration-tests/protractor.conf.ts --suite environment
$ ts-node -O '{"module":"commonjs"}' ./node_modules/.bin/protractor integration-tests/protractor.conf.ts --suite environment
Report destination:   gui_test_screenshots/protractor-report.html
[14:57:14] I/launcher - Running 1 instances of WebDriver
[14:57:14] I/direct - Using ChromeDriver directly...

>> Executing 7 defined specs...

Test Suites & Specs:

1) Create a test namespace
   ✔ creates test namespace if necessary

2) Interacting with the environment variable editor

   3) When a variable is added
      ✔ shows the correct variables

   4) When a variable is deleted
      ✖ does not show any variables (2 failures)

   5) When a variable is added from a config map
      • shows the correct variables

   6) When a variable is deleted from a config map
      • shows the correct variables

   7) When a variable is added from a secret
      • shows the correct variables

   8) When a variable is deleted from a secret
      • shows the correct variables

>> Done!


Failed Specs:

1) Interacting with the environment variable editor When a variable is deleted : does not show any variables
   TimeoutError: Wait timed out after 15004ms
       (removed stack)

   From: Task: <anonymous wait>
       (removed stack)

   From: Task: Run it"does not show any variables" in control flow
       (removed stack)
   
   From asynchronous test: 
       (removed stack)

   Error: Failed: unhandled error: Uncaught Error: Rendered more hooks than during the previous render. Error: Rendered more hooks than during the previous render.
       at updateWorkInProgressHook http://localhost:9000/static/vendors~main-66ab5ad094ae624f5a6e.js:285955:15
       at updateReducer http://localhost:9000/static/vendors~main-66ab5ad094ae624f5a6e.js:286013:14
       at updateState http://localhost:9000/static/vendors~main-66ab5ad094ae624f5a6e.js:286434:10
       at Object.useState http://localhost:9000/static/vendors~main-66ab5ad094ae624f5a6e.js:287235:16
       at useState http://localhost:9000/static/vendors~main-66ab5ad094ae624f5a6e.js:329187:21
       at useTranslation http://localhost:9000/static/vendors~main-66ab5ad094ae624f5a6e.js:300417:73
       at NameKeyDropdownPair http://localhost:9000/static/main-5c800e219ce989ab4a5c.js:152042:75
       at ConfigMapSecretRef http://localhost:9000/static/main-5c800e219ce989ab4a5c.js:152284:10
       at renderWithHooks http://localhost:9000/static/vendors~main-66ab5ad094ae624f5a6e.js:285778:18
       (removed more lines)

Summary:

Suites:  8 of 8
Specs:   3 of 7 (4 disabled)
Expects: 5 (2 failures)
Finished in 32.860 seconds

Last browser URL:  http://localhost:9000/k8s/ns/test-krwov/deployments?name=env-test-krwov
[14:57:59] I/launcher - 0 instance(s) of WebDriver still running
[14:57:59] I/launcher - chrome #01 failed 1 test(s)
[14:57:59] I/launcher - overall: 1 failed spec(s)
Closing report
[14:57:59] E/launcher - Process exited with error code 1
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

**Comment by jerolimov** [comment]
This PR includes now also a fix for this unstable protractor test (and maybe an issue in production code was well):

> Uncaught Error: Rendered more hooks than during the previous render).

`ConfigMapSecretKeyRef` called `NameKeyDropdownPair` as a function instead of rendering it as child component. This is should be avoided since the component is rendered conditionally and uses hooks itself. This was not a problem before the hook for localization (`useTranslation`) was added.

I run `yarn test-protractor-suite --suite environment` multiple times for the main branch and this PR with and without this additional changes to check if this stabilize our integration tests. The result makes me positive:

| variant                                           | failed | successful | overall |
|---------------------------------------------------|--------|------------|---------|
| main branch                                       | 3 (5%) | 57         | 60      |
| PR without changing function call to a compontent | 2 (5%) | 38         | 40      |
| PR with changing function call to a compontent    | 0 🎉   | 40         | 40      |

@nemesis09 Aritra, can you please take a new look and add lgtm if everything is fine. Thanks.

@spadgett Sam, I expect that we want backport this fixes (fix 2 potencial crashes and improved test stability) also to 4.7 and 4.8?

/approve

**Comment by jerolimov** [comment]
Tests are successfully the first time. Force re-run

/test e2e-gcp-console

**Comment by nemesis09** [comment]
verified changes.
e-2-e tests passed. Thanks @jerolimov  for the changes!
/lgtm

**Pull request merged**
